clients/client-go/tcobject/upload.go

```diff
@@ -0,0 +1,160 @@
+package tcobject
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/cenkalti/backoff/v3"
+	"github.com/taskcluster/httpbackoff/v3"
+	"github.com/taskcluster/slugid-go/slugid"
+	tcclient "github.com/taskcluster/taskcluster/v42/clients/client-go"
+)
+
+const (
+	DataInlineMaxSize = 8192
+)
+
+// UploadFromBuf is a convenience method to publish an Object to the Object
+// Server with content buf and given projectID, name, contentType and expires.
+func (object *Object) UploadFromBuf(projectID string, name string, contentType string, expires time.Time, buf []byte) (err error) {
+	readSeeker := bytes.NewReader(buf)
+	contentLength := int64(len(buf))
+	return object.UploadFromReadSeeker(projectID, name, contentType, contentLength, expires, readSeeker)
+}
+
+// UploadFromFile is a convenience method to publish an Object to the Object
+// Server from file with path filepath and given projectID, name, contentType
+// and expires.
+func (object *Object) UploadFromFile(projectID string, name string, contentType string, expires time.Time, filepath string) (err error) {
+	file, err := os.Open(filepath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+	contentLength := fileInfo.Size()
+	return object.UploadFromReadSeeker(projectID, name, contentType, contentLength, expires, file)
+}
+
+// UploadFromReadSeeker publishes an Object to the Object Server, with given
+// name, projectID, contentType, contentLength and expiry, with the object
+// content read from readSeeker. The value of contentLength is not validated
+// prior to upload.
+func (object *Object) UploadFromReadSeeker(projectID string, name string, contentType string, contentLength int64, expires time.Time, readSeeker io.ReadSeeker) (err error) {
+
+	uploadID := slugid.Nice()
+	proposedUploadMethods := ProposedUploadMethods{}
+
+	if contentLength < DataInlineMaxSize {
+		content, err := ioutil.ReadAll(readSeeker)
+		if err != nil {
+			return err
+		}
+		b64 := base64.StdEncoding.EncodeToString(content)
+		proposedUploadMethods.DataInline = DataInlineUploadRequest{
+			ContentType: contentType,
+			ObjectData:  b64,
+		}
+	}
+
+	proposedUploadMethods.PutURL = PutURLUploadRequest{
+		// TODO: find out why this is float64 when it should be int64
```

Comment:
I think this is because it's a "number" in JSON.  Changing the json-schema type to "integer" will probably fix it.

Reply:
Created #4493